internal/oci/container.go

```diff
@@ -347,17 +388,104 @@ func (c *Container) exitFilePath() string {
 	return filepath.Join(c.dir, "exit")
 }
 
-// IsAlive is a function that checks if a container's init pid exists.
+// IsAlive is a function that checks if a container's init PID exists.
 // It is used to check a container state when we don't want a `$runtime state` call
 func (c *Container) IsAlive() bool {
-	if _, err := findprocess.FindProcess(c.state.Pid); err != nil {
-		logrus.Errorf("container %s not running: %v", c.id, err)
+	_, err := c.pid()
+	if err != nil {
+		logrus.Errorf("checking if PID of %s is running failed: %v", c.id, err)
 		return false
 	}
 
 	return true
 }
 
+// Pid returns the container's init PID.
+// It will fail if the saved PID no longer belongs to the container.
+func (c *Container) Pid() (int, error) {
+	c.opLock.Lock()
+	defer c.opLock.Unlock()
+	return c.pid()
+}
+
+// pid returns the container's init PID.
+// It checks that we have an InitPid defined in the state, that PID can be found
+// and it is the same process that was originally started by the runtime.
+func (c *Container) pid() (int, error) {
+	if c.state == nil {
+		return 0, errors.New("state not initialized")
+	}
+	pid := c.state.InitPid
+	if pid <= 0 {
+		return 0, errors.New("PID not initialized")
+	}
+
+	// container has stopped (as pid is initialized but the runc state has overwritten it)
+	if c.state.Pid == 0 {
+		return 0, findprocess.ErrNotFound
+	}
+
+	if err := c.verifyPid(pid); err != nil {
+		return 0, err
+	}
+	return pid, nil
+}
+
+// findAndReleasePid attempts to find the container's PID
+// and if it's found, release it. In between these steps,
+// it also verifies the PID running is the same as was originally
+// started by the runtime.
+func (c *Container) findAndReleasePid() (bool, error) {
+	pid := c.state.InitPid
+
+	if err := c.verifyPid(pid); err != nil {
+		if !errors.Is(err, findprocess.ErrNotFound) {
+			return false, err
+		}
+		return false, nil
+	}
+
+	process, err := findprocess.FindProcess(pid)
+	if err == findprocess.ErrNotFound {
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+
+	// the process was found, and was the correct one; release it
+	if err := process.Release(); err != nil {
+		return true, err
+	}
+	return true, nil
+}
+
+// verifyPid checks that the start time for the process on the node is the same
+// as the start time we saved after creating the container.
+// This is the simplest way to verify we are operating on the container
+// process, and haven't run into PID wrap.
+func (c *Container) verifyPid(pid int) error {
+	startTime, err := getInitStartTime(pid)
+	if err != nil {
+		return err
+	}
+
+	if startTime != c.state.InitStartTime {
+		return errors.New("PID running but not the original container. PID wrap may have occurred")
+	}
+	return nil
+}
+
+// getInitStartTime reads the kernel's /proc entry for stime for PID.
+func getInitStartTime(pid int) (int, error) {
```

Comment:
The word `Init` in the function name is misleading, since the function doesn't care if the `pid` is that of init or not.